src/plugins/ui_actions/public/context_menu/build_eui_context_menu_panels.tsx

```diff
@@ -41,138 +44,178 @@ interface ActionWithContext<Context extends BaseContext = BaseContext> {
+  const morePanelId = panel.id + '__more';
+  const more: ItemDescriptor = {
+    name: txtMore,
+    panel: morePanelId,
+    icon: 'boxesHorizontal',
+    'data-test-subj': `embeddablePanelMore-${id}`,
+    _order: -1,
+  };
+  panel.items = [...visibleItems, more];
+  const subPanel: PanelDescriptor = {
+    id: morePanelId,
+    title: panel.title || defaultTitle,
+    items: itemsInSubmenu,
+  };
+  panels[morePanelId] = subPanel;
+};
+
+const removeItemMetaFields = (items: ItemDescriptor[]): EuiContextMenuPanelItemDescriptor[] => {
+  const euiItems: EuiContextMenuPanelItemDescriptor[] = [];
+  for (const item of items) {
+    const { _order: omit, _title: omit2, ...rest } = item;
+    euiItems.push(rest);
+  }
+  return euiItems;
+};
+
+const removePanelMetaFields = (panels: PanelDescriptor[]): EuiContextMenuPanelDescriptor[] => {
+  const euiPanels: EuiContextMenuPanelDescriptor[] = [];
+  for (const panel of panels) {
+    const { _level: omit, _icon: omit2, ...rest } = panel;
+    euiPanels.push({ ...rest, items: removeItemMetaFields(rest.items) });
+  }
+  return euiPanels;
+};
+
+export interface BuildContextMenuParams {
   actions: ActionWithContext[];
   title?: string;
   closeMenu?: () => void;
-}): Promise<EuiContextMenuPanelDescriptor> {
-  const menuItems = await buildEuiContextMenuPanelItems({
-    actions,
-    closeMenu,
-  });
-
-  return {
-    id: 'mainMenu',
-    title,
-    items: menuItems,
-  };
 }
 
 /**
- * Transform an array of Actions into the shape needed to build an EUIContextMenu
+ * Transforms an array of Actions to the shape EuiContextMenuPanel expects.
  */
-async function buildEuiContextMenuPanelItems({
+export async function buildContextMenuForActions({
   actions,
-  closeMenu,
-}: {
-  actions: ActionWithContext[];
-  closeMenu: () => void;
-}) {
-  actions = sortBy(
-    actions,
-    (a) => -1 * (a.action.order ?? 0),
-    (a) => a.action.type,
-    (a) => a.action.getDisplayName({ ...a.context, trigger: a.trigger })
-  );
-
-  const items: EuiContextMenuPanelItemDescriptor[] = new Array(actions.length);
-  const promises = actions.map(async ({ action, context, trigger }, index) => {
-    const isCompatible = await action.isCompatible({
-      ...context,
-      trigger,
-    });
-    if (!isCompatible) {
-      return;
+  title = defaultTitle,
+  closeMenu = () => {},
+}: BuildContextMenuParams): Promise<EuiContextMenuPanelDescriptor[]> {
+  const panels: Record<string, PanelDescriptor> = {
+    mainMenu: {
+      id: 'mainMenu',
+      title,
+      items: [],
+    },
+  };
+  const promises = actions.map(async (item) => {
+    const { action } = item;
+    const context: ActionExecutionContext<object> = { ...item.context, trigger: item.trigger };
+    const isCompatible = await item.action.isCompatible(context);
+    if (!isCompatible) return;
+    let parentPanel = '';
+    let currentPanel = '';
+    if (action.grouping && action.grouping) {
```

Comment:
`action.grouping && action.grouping` seems like this condition is incorrect?

Reply:
Fixed.